### PR TITLE
Refactor repository API

### PR DESCRIPTION
**Pull request by ywk253100**

### CODE REVIEW

File: src/ui/api/base.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+	"github.com/vmware/harbor/src/common/utils/log"
+	"github.com/vmware/harbor/src/ui/filter"
+	"github.com/vmware/harbor/src/ui/projectmanager"
+)
+
+// BaseController ...
+type BaseController struct {
+	api.BaseAPI
+	// SecurityCxt is the security context used to authN &authZ
+	SecurityCxt security.Context
```
Comment (reasonerjt) [inline src/ui/api/base.go]: SecurityCxt  -> SecurityCtx

File: src/ui/api/repository.go [hunk 2]
```diff
@@ -409,20 +366,20 @@ func (ra *RepositoryAPI) GetManifests() {
...
+		if !ra.SecurityCxt.IsAuthenticated() {
+			ra.HandleUnauthorized()
+			return
+		}
 
-	if project.Public == 0 {
-		userID := ra.ValidateUser()
-		if !checkProjectPermission(userID, project.ProjectID) {
-			ra.CustomAbort(http.StatusForbidden, "")
+		if !ra.SecurityCxt.HasReadPerm(projectName) {
```
Comment (reasonerjt) [inline src/ui/api/repository.go]: Seems HasReadPerm is enough already?

File: src/ui/api/repository.go [hunk 3]
```diff
@@ -68,50 +65,43 @@ type manifestResp struct {
...
 	}
 
-	if project.Public == 0 {
-		var userID int
-
-		if svc_utils.VerifySecret(ra.Ctx.Request, config.JobserviceSecret()) {
-			userID = 1
-		} else {
-			userID = ra.ValidateUser()
+	if !ra.ProManager.IsPublic(projectID) {
```
Comment (reasonerjt) [inline src/ui/api/repository.go]: HasReadPerm should be enough

File: src/ui/api/base.go [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+	// ProManager is the project manager which abstracts the operations
+	// related to projects
+	ProManager projectmanager.ProjectManager
```
Comment (wy65701436) [inline src/ui/api/base.go]: ProjectManager or ProjectMgr is better, I think.

File: src/ui/projectmanager/db/pm.go [hunk 5]
```diff
@@ -108,3 +104,32 @@ func (p *ProjectManager) GetRoles(username string, projectIDOrName interface{})
 
 	return roles
 }
+
+// GetPublic returns all public projects
+func (p *ProjectManager) GetPublic() []models.Project {
```
Comment (wy65701436) [inline src/ui/projectmanager/db/pm.go]: GetPublicProjects() is better?

File: src/ui/projectmanager/db/pm.go [hunk 6]
```diff
@@ -108,3 +104,32 @@ func (p *ProjectManager) GetRoles(username string, projectIDOrName interface{})
...
+	if err != nil {
+		log.Errorf("failed to get all public projects: %v", err)
+		return []models.Project{}
+	}
+
+	return projects
+}
+
+// GetByMember returns all projects which the user is a member of
+func (p *ProjectManager) GetByMember(username string) []models.Project {
```
Comment (wy65701436) [inline src/ui/projectmanager/db/pm.go]: GetProjectByMember() ?

File: src/ui/projectmanager/db/pm.go (hunk 5 above)
Comment (ywk253100) [inline src/ui/projectmanager/db/pm.go]: It is called by ProjectManager.GetPublic(), so I think the prefix may declare the return are projects clearly.

**Pull request merged**